src/perform_step/firk_perform_step.jl

```diff
@@ -749,3 +775,528 @@ end
+    @unpack c1, c2, c3, c4, γ, α1, β1, α2, β2, e1, e2, e3, e4, e5 = cache.tab
+    @unpack κ, cont1, cont2, cont3, cont4 = cache
+    @unpack internalnorm, abstol, reltol, adaptive = integrator.opts
+    alg = unwrap_alg(integrator, true)
+    @unpack maxiters = alg
+    mass_matrix = integrator.f.mass_matrix
+
+    # precalculations
+    rtol = @.. broadcast=false reltol^(2 / 3)/10
+    atol = @.. broadcast=false rtol*(abstol / reltol)
+    c1m1 = c1 - 1
+    c2m1 = c2 - 1
+    c3m1 = c3 - 1
+    c4m1 = c4 - 1
+    c1mc2 = c1 - c2
+    c1mc3 = c1 - c3
+    c1mc4 = c1 - c4
+    c2mc3 = c2 - c3
+    c2mc4 = c2 - c4
+    c3mc4 = c3 - c4
+
+    γdt, α1dt, β1dt, α2dt, β2dt = γ / dt, α1 / dt, β1 / dt, α2 / dt, β2 / dt
+    J = calc_J(integrator, cache)
+    if u isa Number
+        LU1 = -γdt * mass_matrix + J
+        LU2 = -(α1dt + β1dt * im) * mass_matrix + J
+        LU3 = -(α2dt + β2dt * im) * mass_matrix + J
+    else
+        LU1 = lu(-γdt * mass_matrix + J)
+        LU2 = lu(-(α1dt + β1dt * im) * mass_matrix + J)
+        LU3 = lu(-(α2dt + β2dt * im) * mass_matrix + J)
+    end
+    integrator.stats.nw += 1
+
+    # TODO better initial guess
+    if integrator.iter == 1 || integrator.u_modified || alg.extrapolant == :constant
+        cache.dtprev = one(cache.dtprev)
+        z1 = w1 = map(zero, u)
+        z2 = w2 = map(zero, u)
+        z3 = w3 = map(zero, u)
+        z4 = w4 = map(zero, u)
+        z5 = w5 = map(zero, u)
+        cache.cont1 = map(zero, u)
+        cache.cont2 = map(zero, u)
+        cache.cont3 = map(zero, u)
+        cache.cont4 = map(zero, u)
+    else # add cont4
+        c5′ = dt / cache.dtprev
+        c1′ = c1 * c5′
+        c2′ = c2 * c5′
+        c3′ = c3 * c5′
+        c4′ = c4 * c5′
+        z1 = @.. broadcast=false c1′*(cont1 + (c1′ - c3m1) * (cont2 + (c1′ - c2m1) * (cont3 + (c1′ - c1m1) * cont4)))
+        z2 = @.. broadcast=false c2′*(cont1 + (c2′ - c3m1) * (cont2 + (c2′ - c2m1) * (cont3 + (c2′ - c1m1) * cont4)))
+        z3 = @.. broadcast=false c3′*(cont1 + (c3′ - c3m1) * (cont2 + (c3′ - c2m1) * (cont3 + (c3′ - c1m1) * cont4)))
+        z4 = @.. broadcast=false c4′*(cont1 + (c4′ - c3m1) * (cont2 + (c4′ - c2m1) * (cont3 + (c4′ - c1m1) * cont4)))
+        z5 = @.. broadcast=false c5′*(cont1 + (c5′ - c3m1) * (cont2 + (c5′ - c2m1) * (cont3 + (c5′ - c1m1) * cont4)))
+        w1 = @.. broadcast=false TI11*z1 + TI12*z2 + TI13*z3 + TI14*z4 + TI15*z5
+        w2 = @.. broadcast=false TI21*z1 + TI22*z2 + TI23*z3 + TI24*z4 + TI25*z5
+        w3 = @.. broadcast=false TI31*z1 + TI32*z2 + TI33*z3 + TI34*z4 + TI35*z5
+        w4 = @.. broadcast=false TI41*z1 + TI42*z2 + TI43*z3 + TI34*z4 + TI45*z5
+        w5 = @.. broadcast=false TI51*z1 + TI52*z2 + TI53*z3 + TI54*z4 + TI55*z5
+    end
+
+    # Newton iteration
+    local ndw
+    η = max(cache.ηold, eps(eltype(integrator.opts.reltol)))^(0.8)
+    fail_convergence = true
+    iter = 0
+    while iter < maxiters
```

Comment:
Why are you writing out newton iterations yourself? Shouldn't this be using nlsolve like all the other implicit methods?

Reply:
oh, I see that all the Radau methods are like this. @ChrisRackauckas why are these using hand rolled newton solvers?

Reply:
because it has to do a very specific FIRK newton which doesn't match the form of the others.

Reply:
As a general TODO, we should also move these to NonlinearSolve.jl, but that is separate from this PR.